File: hector_pipe_detection/src/edge_detection.cpp
//
// Created by Stefan Fabian on 12.06.18.
//

#include "hector_pipe_detection/edge_detection.h"

#include <opencv2/opencv.hpp>

namespace hector_pipe_detection
{

template<typename T>
void calculateThresholds( const cv::Mat &mat, double &upper, double &lower )
{
  double old_upper = 0;
  double old_lower = 0;
  upper = 160;
  lower = 80;
  while ( cv::abs( upper - old_upper ) > 0.1 || cv::abs( lower - old_lower ) > 0.1 )
  {
    old_upper = upper;
    old_lower = lower;
    int count_upper = 0;
    int count_middle = 0;
    int count_lower = 0;
    double upper_sum = 0;
    double middle_sum = 0;
    double lower_sum = 0;
    for ( int row = 0; row < mat.rows; ++row )
    {
      for ( int col = 0; col < mat.cols; ++col )
      {
        T val = mat.at<T>( row, col );
        if ( val > upper )
        {
          ++count_upper;
          upper_sum += val;
        }
        else if ( val > lower )
        {
          ++count_middle;
          middle_sum += val;
        }
        else
        {
          ++count_lower;
          lower_sum += val;
        }
      }
    }
    upper = (upper_sum / count_upper + middle_sum / count_middle) / 2;
    lower = (middle_sum / count_middle + lower_sum / count_lower) / 2;
  }
//  std::cout << "Lower threshold: " << lower << ", Upper: " << upper << std::endl;
}

void applyFilterAndMaxChannel( const cv::Mat &mat, cv::Mat &out, const cv::Mat &filter )
{
  cv::Mat channels[3];
  cv::Mat response;
  cv::filter2D( mat, response, -1, filter );
  cv::split( cv::abs( response ), channels );
  out = cv::max( channels[0], cv::max( channels[1], channels[2] ));
}

void combineResponses( cv::Mat &out, const cv::Mat &in, cv::Mat &orientation, const char orientation_in )
{
  for ( int row = 0; row < out.rows; ++row )
  {
    for ( int col = 0; col < out.cols; ++col )
    {
      if ( out.at<short>( row, col ) > in.at<short>( row, col )) continue;
      out.at<short>( row, col ) = in.at<short>( row, col );
      orientation.at<char>( row, col ) = orientation_in;
    }
  }
}

template<typename T>
inline bool isSmaller( const cv::Mat &mat, int x1, int y1, int x2, int y2 )
{
  if ( x2 < 0 || x2 >= mat.cols ) return false;
  if ( y2 < 0 || y2 >= mat.rows ) return false;
  return mat.at<T>( y1, x1 ) <= mat.at<T>( y2, x2 );
}

void cdm( const cv::Mat &image, cv::Mat &out )
{
  cv::Mat mean_filtered;
  cv::boxFilter( image, mean_filtered, CV_16S, cv::Size( 5, 5 ));

  cv::Mat diagonal_tl_to_br( 3, 3, CV_16S, cv::Scalar( 0 ));
  diagonal_tl_to_br.at<short>( 0, 0 ) = 1;
  diagonal_tl_to_br.at<short>( 2, 2 ) = -1;

  cv::Mat vertical_filter( 3, 1, CV_16S, cv::Scalar( 0 ));
  vertical_filter.at<short>( 0, 0 ) = 1;
  vertical_filter.at<short>( 2, 0 ) = -1;

  cv::Mat diagonal_tr_to_bl( 3, 3, CV_16S, cv::Scalar( 0 ));
  diagonal_tr_to_bl.at<short>( 0, 2 ) = 1;
  diagonal_tr_to_bl.at<short>( 2, 0 ) = -1;

  cv::Mat horizontal_filter( 1, 3, CV_16S, cv::Scalar( 0 ));
  horizontal_filter.at<short>( 0, 0 ) = 1;
  horizontal_filter.at<short>( 0, 2 ) = -1;

  cv::Mat max_response;
  applyFilterAndMaxChannel( mean_filtered, max_response, diagonal_tl_to_br );
  // Orientations: 0: TL2BR, 1: Vertical, 2: BL2TR, 3: Horizontal
  cv::Mat orientation( max_response.rows, max_response.cols, CV_8S, cv::Scalar( 0 ));

  cv::Mat response;
  applyFilterAndMaxChannel( mean_filtered, response, vertical_filter );
  combineResponses( max_response, response, orientation, 1 );

  applyFilterAndMaxChannel( mean_filtered, response, diagonal_tr_to_bl );
  combineResponses( max_response, response, orientation, 2 );

  applyFilterAndMaxChannel( mean_filtered, response, horizontal_filter );
  combineResponses( max_response, response, orientation, 3 );

  cv::Mat mean, stddev;
  cv::meanStdDev( max_response, mean, stddev );

  out = cv::Mat( max_response.rows, max_response.cols, CV_8U, cv::Scalar( 0 ));
  for ( int row = 0; row < max_response.rows; ++row )
  {
    for ( int col = 0; col < max_response.cols; ++col )
    {
      if ( max_response.at<short>( row, col ) < mean.at<double>( 0, 0 ))
      {
        continue;
      }
      bool zero;
      switch ( orientation.at<char>( row, col ))
      {
        case 0:
          zero = isSmaller<short>( max_response, col, row, col - 1, row + 1 ) ||
                 isSmaller<short>( max_response, col, row, col + 1, row - 1 );
          break;
        case 1:
          zero = isSmaller<short>( max_response, col, row, col, row + 1 ) ||
                 isSmaller<short>( max_response, col, row, col, row - 1 );
          break;
        case 2:
          zero = isSmaller<short>( max_response, col, row, col - 1, row - 1 ) ||
                 isSmaller<short>( max_response, col, row, col + 1, row + 1 );
          break;
        case 3:
          zero = isSmaller<short>( max_response, col, row, col - 1, row ) ||
                 isSmaller<short>( max_response, col, row, col + 1, row );
          break;
        default:
          zero = true;
      }
      if ( zero ) continue;
      out.at<unsigned char>( row, col ) = 255;
    }
  }

//  cv::threshold( max_response, max_response, mean.at<double>( 0, 0 ) + stddev.at<double>( 0, 0 ), 255,
//                 CV_THRESH_TOZERO );
//  max_response.convertTo( out, CV_8U );
}

void getJacobians( const cv::Mat &response_x, const cv::Mat &response_y, cv::Mat &jx, cv::Mat &jy, cv::Mat &jxy )
{
  cv::Mat channels_x[3];
  cv::split( response_x, channels_x );
  jx = channels_x[0].mul( channels_x[0] );
  jx += channels_x[1].mul( channels_x[1] );
  jx += channels_x[2].mul( channels_x[2] );

  cv::Mat channels_y[3];
  cv::split( response_y, channels_y );
  jy = channels_y[0].mul( channels_y[0] );
  jy += channels_y[1].mul( channels_y[1] );
  jy += channels_y[2].mul( channels_y[2] );

  jxy = channels_x[0].mul( channels_y[0] );
  jxy += channels_x[1].mul( channels_y[1] );
  jxy += channels_x[2].mul( channels_y[2] );
}

template<typename T>
inline void flowThreshold(const cv::Mat &edges, cv::Mat &binary, double lower, int row, int col)
{
  for (int i = -1; i <= 1; ++i)
  {
    for (int j = -1; j <= 1; ++j)
    {
      if (i == 0 && j == 0) continue;
      int r = row + i;
      int c = col + j;
      if (r < 0 || r == edges.rows) continue;
      if (c < 0 || c == edges.cols) continue;
      if (binary.at<unsigned char>(r, c) == 255 || edges.at<T>(r, c) < lower) continue;
      binary.at<unsigned char>(r, c) = 255;
      flowThreshold<T>(edges, binary, lower, r, c);
    }
  }
}

template<typename T>
void threshold(const cv::Mat &edges, cv::Mat &binary, double upper, double lower)
{
  binary = cv::Mat(edges.rows, edges.cols, CV_8U, cv::Scalar(0));
  for (int row = 0; row < edges.rows; ++row)
  {
    for (int col = 0; col < edges.cols; ++col)
    {
      if ( edges.at<T>( row, col ) >= upper && binary.at<unsigned char>( row, col ) != 255 )
      {
        binary.at<unsigned char>( row, col ) = 255;
        flowThreshold<T>( edges, binary, lower, row, col );
      }
    }
  }
}

void color_edges( const cv::Mat &image, cv::Mat &out )
{
  int filter_type = CV_16S;
  cv::Mat x_filter( 3, 3, filter_type, cv::Scalar( 0 ));
  typedef short FilterType;
  x_filter.at<FilterType>( 0, 0 ) = 1;
  x_filter.at<FilterType>( 0, 1 ) = 2;
  x_filter.at<FilterType>( 0, 2 ) = 1;
  x_filter.at<FilterType>( 2, 0 ) = -1;
  x_filter.at<FilterType>( 2, 1 ) = -2;
  x_filter.at<FilterType>( 2, 2 ) = -1;
  cv::Mat y_filter;
  cv::transpose( x_filter, y_filter );

  cv::Mat response_x;
  cv::filter2D( image, response_x, filter_type, x_filter );
  response_x.convertTo( response_x, CV_32F );

  cv::Mat response_y;
  cv::filter2D( image, response_y, filter_type, y_filter );
  response_y.convertTo( response_y, CV_32F );

  cv::Mat jx, jy, jxy;
  getJacobians( response_x, response_y, jx, jy, jxy );

  // compute first (greatest) eigenvalue of 2x2 matrix J'*J.
  // note that the abs() is only needed because some values may be slightly
  // negative due to round-off error.
  cv::Mat d;
  cv::sqrt( cv::abs( jx.mul( jx ) - 2 * jx.mul( jy ) + jy.mul( jy ) + 4 * jxy.mul( jxy )), d );
  cv::Mat e1 = (jx + jy + d) / 2;
  // the 2nd eigenvalue would be:  e2 = (Jx + Jy - D) / 2
  cv::Mat edges;
  cv::sqrt( e1, edges );
  double upper;
  double lower;
  calculateThresholds<float>( edges, upper, lower );
  threshold<float>(edges, out, upper, lower);
  //cv::threshold( edges, out, upper, 255, CV_THRESH_BINARY );
  out.convertTo( out, CV_8U );
  return;

  /*cv::Mat mean, stddev;
  cv::meanStdDev( edges, mean, stddev );
  double high_threshold = mean.at<double>( 0, 0 ) + 2 * stddev.at<double>( 0, 0 );
  double low_threshold = mean.at<double>( 0, 0 ) + 0.5 * stddev.at<double>( 0, 0 );
  out = cv::Mat( edges.rows, edges.cols, CV_8U );
  float min_orientation = 360;
  float max_orientation = 0;

  for ( int row = 0; row < edges.rows; ++row )
  {
    for ( int col = 0; col < edges.cols; ++col )
    {
      // 0 degrees is the positive X-Axis, degrees increment counter-clockwise
      // We move it so that up is 90, down is -90 and flip the left side to the right
      float orientation =
        cv::fastAtan2( -jxy.at<float>( row, col ), e1.at<float>( row, col ) - jy.at<float>( row, col )) - 90;
      if ( orientation > 270 ) orientation -= 360;
      else if ( orientation > 90 ) orientation -= 180;
      if ( orientation > max_orientation ) max_orientation = orientation;
      if ( orientation < min_orientation ) min_orientation = orientation;
      bool zero;
      if ( edges.at<float>( row, col ) < high_threshold )
      {
        if ( edges.at<float>( row, col ) >= low_threshold )
        {
          out.at<char>( row, col ) = 127;
          continue;
        }
        zero = true;
      }
      else if ( orientation > 67.5 || orientation <= -67.5 ) // Vertical
      {
        zero = isSmaller<float>( edges, col, row, col, row - 1 ) || isSmaller<float>( edges, col, row, col, row + 1 );
      }
      else if ( orientation > 22.5 ) // Bottom-Left to Top-Right
      {
        zero = isSmaller<float>( edges, col, row, col + 1, row - 1 ) ||
               isSmaller<float>( edges, col, row, col - 1, row + 1 );
      }
      else if ( orientation > -22.5 ) // Horizontal
      {
        zero = isSmaller<float>( edges, col, row, col - 1, row ) || isSmaller<float>( edges, col, row, col + 1, row );
      }
      else // Top-Left to Bottom-Right
      {
        zero = isSmaller<float>( edges, col, row, col - 1, row - 1 ) ||
               isSmaller<float>( edges, col, row, col + 1, row + 1 );
      }

      out.at<char>( row, col ) = zero ? (char) 0 : (char) 255;
    }
  }*/
}
}

File: hector_fast_hazmat_detection/src/edge_detection.py
import cv2
import numpy as np


def edgedetect (channel):
    sobelX = cv2.Sobel(channel, cv2.CV_16S, 1, 0)
    sobelY = cv2.Sobel(channel, cv2.CV_16S, 0, 1)
    sobel = np.hypot(sobelX, sobelY)

    sobel[sobel > 255] = 255  # Some values seem to go above 255. However RGB channels has to be within 0-255
    return sobel


def cdm(image, image_mem, threshold=None):
    mean_filtered = cv2.filter2D(image, cv2.CV_16S, np.ones((5, 5)) / 25)
    filters = np.array([[[1, 0, 0],
                         [0, 0, 0],
                         [0, 0, -1]],
                        [[0, 1, 0],
                         [0, 0, 0],
                         [0, -1, 0]],
                        [[0, 0, 1],
                         [0, 0, 0],
                         [-1, 0, 0]],
                        [[0, 0, 0],
                         [1, 0, -1],
                         [0, 0, 0]]])
    max_response = np.zeros((image_mem.shape[0], image_mem.shape[1]))
    for i in range(len(filters)):
        response = cv2.filter2D(mean_filtered, cv2.CV_16S, filters[i, :, :])
        response = np.max(np.abs(response.get() if isinstance(response, cv2.UMat) else response), axis=2)
        max_response = np.maximum(response, max_response)
    if threshold is None:
        mean = np.mean(max_response[max_response > 0])
        std = np.var(max_response[max_response > 0])
        threshold = mean # + 0.05 * std
    _, max_response = cv2.threshold(max_response, threshold, 255, cv2.CV_8U)
    return max_response


def color_edges(image):
    im = image.astype(np.float) / 255

    y_filter = np.array([[1, 2, 1], [0, 0, 0], [-1, -2, -1]])
    x_filter = y_filter.T

    rx = cv2.filter2D(image[:, :, 0], cv2.CV_32F, x_filter)
    gx = cv2.filter2D(image[:, :, 1], cv2.CV_32F, x_filter)
    bx = cv2.filter2D(image[:, :, 2], cv2.CV_32F, x_filter)

    ry = cv2.filter2D(image[:, :, 0], cv2.CV_32F, y_filter)
    gy = cv2.filter2D(image[:, :, 1], cv2.CV_32F, y_filter)
    by = cv2.filter2D(image[:, :, 2], cv2.CV_32F, y_filter)

    Jx = rx ** 2 + gx ** 2 + bx ** 2
    Jy = ry ** 2 + gy ** 2 + by ** 2
    Jxy = rx * ry + gx * gy + bx * by

    # compute first (greatest) eigenvalue of 2x2 matrix J'*J.
    # note that the abs() is only needed because some values may be slightly
    # negative due to round-off error.
    D = np.sqrt(np.abs(Jx ** 2 - 2 * Jx * Jy + Jy ** 2 + 4 * Jxy ** 2))
    e1 = (Jx + Jy + D) / 2
    # the 2nd eigenvalue would be:  e2 = (Jx + Jy - D) / 2

    edge_magnitude = np.sqrt(e1)

    # compute edge orientation (from eigenvector tangent)
    edge_orientation = np.arctan2(-Jxy, e1 - Jy)
    edge_orientation = edge_orientation * 180 / np.pi
    return (edge_magnitude * 255 / np.max(edge_magnitude)).astype(np.uint8), edge_orientation


def is_smaller(img, x1, y1, x2, y2):
    if x2 < 0 or x2 >= img.shape[0]:
        return False
    if y2 < 0 or y2 >= img.shape[1]:
        return False
    return img[x1, y1] < img[x2, y2]


def non_max_suppression(edges, orientation):
    out = np.empty_like(edges)
    for i in range(edges.shape[0]):
        for j in range(edges.shape[1]):
            if orientation[i, j] > 62.5 or orientation[i, j] < -62.5:
                zero = is_smaller(edges, i, j, i, j - 1) or is_smaller(edges, i, j, i, j + 1)
            elif orientation[i, j] > 22.5:
                zero = is_smaller(edges, i, j, i + 1, j + 1) or is_smaller(edges, i, j, i - 1, j - 1)
            elif orientation[i, j] > -22.5:
                zero = is_smaller(edges, i, j, i + 1, j) or is_smaller(edges, i, j, i - 1, j)
            else:
                zero = is_smaller(edges, i, j, i - 1, j + 1) or is_smaller(edges, i, j, i + 1, j - 1)
            out[i, j] = 0 if zero else edges[i, j]
    return out


def calculate_threshold(edges):
    oldmean = 0
    mean = np.mean(edges)
    while np.abs(mean - oldmean) > 0.1:
        oldmean = mean
        high = np.mean(edges[edges >= mean])
        low = np.mean(edges[edges < mean])
        mean = (high + low) / 2
    return mean


def calculate_thresholds(edges):
    oldupper = 0
    oldlower = 0
    upper = 160
    lower = 80
    while np.abs(upper - oldupper) > 0.1 or np.abs(lower - oldlower) > 0.1:
        oldupper = upper
        oldlower = lower
        high_mean = np.mean(edges[edges > upper])
        middle_mean = np.mean(edges[np.logical_and(lower < edges, edges <= upper)])
        low_mean = np.mean(edges[edges <= lower])

        upper = (high_mean + middle_mean) / 2
        lower = (middle_mean + low_mean) / 2
    return upper, lower


# def flowThreshold(edges, out, lower, row, col):


def threshold(edges, upper, lower):
    highs = []
    indicator = np.zeros_like(edges, dtype=np.uint8)
    for i in range(edges.shape[0]):
        for j in range(edges.shape[1]):
            if edges[i, j] >= upper:
                highs.append((i, j))
                indicator[i, j] = 255
    while len(highs) > 0:
        pt = highs[-1]
        del highs[-1]
        for i in [-1, 0, 1]:
            for j in [-1, 0, 1]:
                if i == 0 and j == 0:
                    continue
                row = pt[0] + i
                col = pt[1] + j
                if row == -1 or row == edges.shape[0] or col == -1 or col == edges.shape[1]:
                    continue
                if indicator[row, col] != 0:
                    continue
                if edges[row, col] < lower:
                    edges[row, col] = 0
                    continue
                indicator[row, col] = 255
                highs.append((row, col))
    #edges[np.logical_and(indicator == 0, edges < upper)] = 0
    return indicator


def extract_foreground_mask(image):
    gpu = isinstance(image, cv2.UMat)
    image_lab = cv2.cvtColor(image, cv2.COLOR_RGB2Lab)
    image_lab_mem = image_lab.get() if gpu else image_lab
    mean = np.mean(image_lab_mem.reshape((image_lab_mem.shape[0] * image_lab_mem.shape[1], 3)), axis=0)
    diff = image_lab_mem - mean
    diff = np.sum(diff[:, :, 1:] ** 2, axis=2)
    diff = diff * 255 / np.max(diff)
    diff = np.round(diff)
    diff[diff < np.mean(diff)] = 0
    diff[diff > 0] = 255
    diff = diff.astype(np.uint8)
    if gpu:
        diff = cv2.UMat(diff)

    diff = cv2.dilate(cv2.erode(diff, None), None, iterations=2)
    return diff


def lab_edge_detection(image):
    image_lab = cv2.cvtColor(image, cv2.COLOR_RGB2LAB)
    image_lab = image_lab.get() if isinstance(image_lab, cv2.UMat) else image_lab

    n = 65
    big_filter = np.ones((n, n))
    big_filter = -big_filter
    big_filter[n // 2, n // 2] = n * n - 1

    # Remove L channel
    response = cv2.filter2D(image_lab[:, :, 1:], cv2.CV_32F, big_filter)
    binary = np.sum(response**2, axis=2)
    binary = binary * 255 / np.max(binary)
    binary = binary.astype(np.uint8)
    mean = cv2.mean(binary)[0]
    _, binary = cv2.threshold(binary, mean, 255, cv2.THRESH_BINARY)
    binary = cv2.dilate(cv2.erode(binary, None), None)
    edges = cv2.Canny(binary, 20, 1)
    return edges
